lib/ukstore/store.c

```diff
@@ -29,3 +29,1194 @@
+	} while (0)
+
+/* Unsigned input, signed etype */
+#define GETCASE_DOWNCASTUS(entry, etype, PTYPE, var, eparam)	\
+	do {							\
+		case uk_store_get_entry_type(etype): {		\
+			__ ## etype val;			\
+								\
+			ret = e->get.etype(&val);		\
+			if (ret < 0)				\
+				return ret;			\
+			if (unlikely(val > __ ## PTYPE ## _MAX))\
+				return -ERANGE;			\
+			*(var)  = (__ ## eparam) val;		\
+			return ret;				\
+		}						\
+	} while (0)
+
+/* Both signed */
+#define GETCASE_DOWNCASTSS(entry, etype, PTYPE, var, eparam)	\
+	do {							\
+		case uk_store_get_entry_type(etype): {		\
+			__ ## etype val;			\
+								\
+			ret = e->get.etype(&val);		\
+			if (ret < 0)				\
+				return ret;			\
+			if (unlikely(val < __ ## PTYPE ## _MIN	\
+				|| val > __ ## PTYPE ##_MAX))	\
+				return -ERANGE;			\
+			*(var)  = (__ ## eparam) val;		\
+			return ret;				\
+		}						\
+	} while (0)
+
+/* Both unsigned */
+#define GETCASE_DOWNCASTUU(entry, etype, PTYPE, var, eparam)	\
+	do {							\
+		case uk_store_get_entry_type(etype): {		\
+			__ ## etype val;			\
+								\
+			ret = e->get.etype(&val);		\
+			if (ret < 0)				\
+				return ret;			\
+			if (unlikely(val > __ ## PTYPE ## _MAX))\
+				return -ERANGE;			\
+			*(var)  = (__ ## eparam) val;		\
+			return ret;				\
+		}						\
+	} while (0)
+
+/* All other cases */
+#define GETCASE_UPCAST(entry, etype, PTYPE, var, eparam)	\
+	do {							\
+		case uk_store_get_entry_type(etype): {		\
+			__ ## etype val;			\
+								\
+			ret = (entry)->get.etype(&val);		\
+			if (ret < 0)				\
+				return ret;			\
+			*(var)  = (__ ## eparam) val;		\
+			return ret;				\
+		}						\
+	} while (0)
+
+/* Unsigned input, signed etype */
+#define GETCASE_UPCASTSU(entry, etype, PTYPE, var, eparam)	\
+	do {							\
+		case uk_store_get_entry_type(etype): {		\
+			__ ## etype val;			\
+								\
+			ret = (entry)->get.etype(&val);		\
+			if (ret < 0)				\
+				return ret;			\
+			if (unlikely(val < 0))			\
+				return -ERANGE;			\
+			*(var)  = (__ ## eparam) val;		\
+			return ret;				\
+		}						\
+	} while (0)
+
+/**
+ * All getters below use this description.
+ * Gets a new value using the save getter and checks the ranges of the values.
+ *
+ * @param e the entry to call the getter from
+ * @param out the value to get in
+ * @return the return value of the getter
+ */
+
+int
+_uk_store_get_u8(struct uk_store_entry *e, __u8 *out)
+{
+	int ret;
+
+	UK_ASSERT(e);
+
+	switch (e->type) {
+	GETCASE_UPCAST(e, u8, U8, out, u8);
+	GETCASE_UPCASTSU(e, s8, U8, out, u8);
+	GETCASE_DOWNCASTUU(e, u16, U8, out, u8);
+	GETCASE_DOWNCASTSU(e, s16, U8, out, u8);
+	GETCASE_DOWNCASTUU(e, u32, U8, out, u8);
+	GETCASE_DOWNCASTSU(e, s32, U8, out, u8);
+	GETCASE_DOWNCASTUU(e, u64, U8, out, u8);
+	GETCASE_DOWNCASTSU(e, s64, U8, out, u8);
+	GETCASE_DOWNCASTUU(e, uptr, U8, out, u8);
+
+	case uk_store_get_entry_type(charp): {
+		char *val = NULL;
+		__s64 nr;
+
+		ret = e->get.charp(&val);
+		if (ret < 0)
+			return ret;
+		nr = atoi(val);
```

Comment:
Can you use `sscanf` and the appropriate`__SCN*` defintions? An s16 should be also enough to detect an ERANGE.